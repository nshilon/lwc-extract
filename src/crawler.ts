import * as Lint from 'tslint'
import * as path from 'path'
import {EOL} from 'os'
import * as fs from 'fs';

import {consoleTestResultHandler} from "tslint/lib/test";


export function crawl({projectDir = './', out = ''} = {}): string {

    let tsConfigPath = path.join(projectDir, 'tsconfig.json');

    const options = {
        fix: false,
        rulesDirectory: path.join(__dirname, 'rules'),
        formattersDirectory: path.join(__dirname, 'formatters'),
        formatter: 'lwc-result',
        allowJS: true
    }
    const program = Lint.Linter.createProgram(tsConfigPath, projectDir)
    const linter = new Lint.Linter(options, program)
    const rules = new Map<string, Partial<Lint.IOptions>>([
        [
            'lwc-extract',
            {
                ruleName: 'lwc-extract',
            },
        ],
    ])
    const lintConfiguration = {
        rules,
        jsRules: rules,
        rulesDirectory: [options.rulesDirectory],
        formattersDirectory: [options.formattersDirectory],
        formatter: 'lwc-result',
        extends: [''],
    }

    const files = Lint.Linter.getFileNames(program)

    files.forEach(file => {
        const fileContents = program.getSourceFile(file).getFullText()
        linter.lint(file, fileContents, lintConfiguration)
    })

    const result = linter.getResult().output;


    if (out) {
        fs.writeFileSync(path.resolve(out), result);
    } else {
        console.log(result);
    }

    return result;

}
