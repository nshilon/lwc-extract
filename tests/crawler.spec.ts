import {crawl} from '../src/crawler'

test(`test extract`, () => {

    const projectDir = './tests/fixtures';
    const result = crawl({projectDir})

    expect(result).toMatchSnapshot();
})
