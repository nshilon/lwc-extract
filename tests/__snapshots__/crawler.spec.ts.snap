// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test extract 1`] = `"[{\\"file\\":\\"propFilter.ts\\",\\"imports\\":{\\"'lwc'\\":[\\"LightningElement\\",\\"wire\\",\\"api\\",\\"track\\"],\\"'lightning/messageService'\\":[\\"publish\\",\\"MessageContext\\"],\\"'@salesforce/messageChannel/FiltersChange__c'\\":[\\"FILTERSCHANGEMC\\"]},\\"exportedClass\\":{\\"PropertyFilter\\":{\\"extends\\":[\\"LightningElement\\"],\\"props\\":[\\"@track searchKey\\",\\"@api minBedrooms\\",\\"@api minBathrooms\\",\\"@wire(MessageContext) messageContext\\"]}}},{\\"file\\":\\"propTile.ts\\",\\"imports\\":{\\"'lwc'\\":[\\"LightningElement\\",\\"api\\"]},\\"exportedClass\\":{\\"PropertyTile\\":{\\"extends\\":[\\"LightningElement\\"],\\"props\\":[\\"@api property\\"]}}}]"`;
